### ISSUE TITLE
Migrate from gost to gb

### ISSUE THREAD

**Issue by xiam** [getlantern/lantern#2595]
Our [gost](https://github.com/getlantern/gost) tool is currently receiving no attention. In the other hand [gb](http://getgb.io/) is getting a lot of traction so it sounds like a good option for us.

See http://getgb.io/examples/getting-started/ and 
https://www.youtube.com/watch?v=c3dW80eO88I

**Comment by uaalto** [comment]
This might be a good issue to get my hands into, even though is P1 is not actually touching functionality and will help me get more familiar with the different sub-projects. What do you think?

**Comment by xiam** [comment]
Hey @uaalto,

I think it would be awesome to have this sorted out! I believe we have some repositories that we don't really need so I think it would be better to try to to build github.com/getlantern/flashlight from scratch and add repos as they're needed. Please be aware that some of the repositories mirrored under `src` may actually differ from their upstream repositories, it will be super useful to identify them so we can fix upstream too.

Thanks,

**Comment by uaalto** [comment]
Roger that! I'll take care, and get back to you if any question arises.

**Comment by aranhoide** [comment]
Note also that some repos may not be needed by flashlight, but we want them
in sync with the flashlight project.  That is the case for the peerscanner
project at least.  Still, I think @xiam's approach is sound to get the port
started.  We can always add whatever projects we miss later.
O 5 xuñ, 2015 18:13, "Ulysses Aalto" <notifications@github.com> escribiu:

**Comment by myleshorton** [comment]
Yeah to echo @xiam's point, at this point the canonical version of those repositories are the ones in https://github.com/getlantern/lantern. To be super clear, for example: 

https://github.com/getlantern/lantern/tree/valencia/src/github.com/getlantern/fronted 

is much more up-to-date than: 

https://github.com/getlantern/fronted

Probably the way to go would be to first push all changes like that upstream (see https://github.com/getlantern/gost#contribute-changes-back-upstream) and to them start from scratch building flashlight alone, theoretically from flashlight itself with its changes pushed upstream, so from https://github.com/getlantern/flashlight, which is very out of date currently.

Does that make sense? It's basically a fairly significant project to undertake =) @uaalto, but have at it if you like! Just getting all the upstream versions updated and compiling would be significant step.

**Comment by uaalto** [comment]
Ok, I have everything lined up to start the migration. There is just one issue that I wanted to checkout with you guys.

The objective of _vendoring_ is to have reproducible builds (i.e. exact versioning of all dependencies, one way being bundling them in the same respository). That seems to be addressed by Gb. What is not so clear to me is that this tool helps solving the upstream synchronization problem. That is, 'gb vendor' plugin (which is actually out of the core scope of Gb) will fetch deps, but won't take care of updating upstream. On the other hand, Gost has this functionality to update upstream repositories.

The problem is that once Gb migration is in place, there won't be an easy way to update upstream repositories without the proper strategy or support. In other words, with Gb you get reproducible builds but at the cost of no default means of updating upstream repositories. Possible solutions would be:

1) A Gb plugin
2) A script using git subtrees or submodules
3) Manually using git subtrees or submodules

What do you think about this? I might be missing something!

**Comment by fffw** [comment]
I think the ability to push back changes to upstream is important for us,
though not a strong requirement for a build tool in general.

gb requires us to manually place all dependencies to a separate `vendor`
folder, even when it's another repo written by ourselves, which is a bit
counterintuitive.

On Mon, Jun 8, 2015 at 10:03 PM, Ulysses Aalto <notifications@github.com>
wrote:

> Ok, I have everything lined up to start the migration. There is just one
> issue that I wanted to checkout with you guys.
>
> The objective of *vendoring* is to have reproducible builds (i.e. exact
> versioning of all dependencies, one way being bundling them in the same
> respository). That seems to be addressed by Gb. What is not so clear to me
> is that this tool helps solving the upstream synchronization problem. That
> is, 'gb vendor' plugin (which is actually out of the core scope of Gb) will
> fetch deps, but won't take care of updating upstream. On the other hand,
> Gost has this functionality to update upstream repositories.
>
> The problem is that once Gb migration is in place, there won't be an easy
> way to update upstream repositories without the proper strategy or support.
> In other words, with Gb you get reproducible builds but at the cost of no
> default means of updating upstream repositories. Possible solutions would
> be:
>
> 1) A Gb plugin
> 2) A script using git subtrees or submodules
> 3) Manually using git subtrees or submodules
>
> What do you think about this? I might be missing something!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/issues/2595#issuecomment-110006389>
> .
>



-- 
BR,
Merlin

**Comment by xiam** [comment]
I agree with @fffw, I think it will require us a bit more effort to push packages to upstream but it won't be impossible. Manually using git submodules sounds to me like an option we can try, if we see this task is repetitive and can be scripted then we could create an script for that.

**Comment by uaalto** [comment]
Yes, I agree that implementing this functionality is better after we've gained some experience. Not implement it until it's needed (YAGNI).
There is a remaining point of discussion: submodules vs subtrees.
Basically, it boils down to (in our use case, as the list is extensive):

Submodules:
* Pros: Dependencies are easily kept in sync. The host project always knows when the submodule needs update or has uncommitted files.
* Cons: Updates must be done _per project_. You can push recursively, but you cannot (AFAIK) link all commits into one so it can be reviewed as a single commit.

Subtrees:
* Pros: Commits include modified files from all dependencies, which simplifies code review and makes commits more consistent.
* Cons: It's more difficult to keep in sync with upstream, and complicates management a little bit.

Gost was using Subtrees, and in my opinion for good reason. It is in any case what would change less how things are working now. Basically, we'd have a vendoring system (Gb) and we would handle subtrees manually (until and iff we see it adequate, we could try any of the options I mentioned earlier). Gost was trying to do both.
What do you think?

**Comment by uaalto** [comment]
I'd add also to the list of options to simplify pushing upstream the use of Git client-side hooks. This could be an easy way of automating upstream updates after a push. I need to see if is viable though.

**Comment by myleshorton** [comment]
Honestly @uaalto I would just start out with the vanilla approach of treating all repos outside our main repo as vendored from the main repo and not as submodules or subtrees but just as vendored code at a given snapshot. If we make changes and push them, we should be able to fairly easily script the update to the upstream repos after a push in a variety of ways, but I like your client-side hook idea.

**Comment by uaalto** [comment]
The process would be:
1) Update upstream repos
2) Rebuild flashlight project, in getlantern/flashlight
3) Make a copy of the dependencies via 'gb vendor', which will pull dependencies without any gitmodule/gitsubtree at all.

Is this correct?

I think you can just push changes upstream afterwards squashing commits anyway, using subtrees.

**Comment by myleshorton** [comment]
That's my understanding of how the vanilla gb approach would work @uaalto, but I haven't looked into it very deeply at all. If subtrees would be transparent to gb and would make pushing upstream easier, then by all means let's use 'em!

**Comment by uaalto** [comment]
Yes, Gb doesn't care at all. It decouples vendoring from version control, and focuses only on the first.

Actually, I think it would be a good idea to use subtrees from the beginning, because they don't add usage overhead, are easy to set up and need to know about them only when pushing upstream. They are mostly like having a local copy that you can push upstream at will, but you are forced to do so. Is basically the same workflow we have currently.

**Comment by uaalto** [comment]
There is news affecting this issue: https://groups.google.com/forum/#!msg/golang-dev/74zjMON9glU/4lWCRDCRZg0J

Apparently, experimental support for vendoring in Go 1.5 will have pretty much the same philosophy of Gb. There seems to be a difference regarding placement of the project source code:

Gb: 
```
.
├── src
│   └── github.com
│       └── pkg
│           └── sftp
│               └── examples
│                   ├── buffered-read-benchmark
│                   ├── buffered-write-benchmark
│                   ├── gsftp
│                   ├── streaming-read-benchmark
│                   └── streaming-write-benchmark
└── vendor
    └── src
        ├── github.com
        │   └── kr
        │       └── fs
        └── golang.org
            └── x
                └── crypto
                    ...
```

go -vendor:

```
$GOPATH
|	src/
|	|	github.com/constabulary/example-gsftp/
|	|	|	cmd/
|	|	|	|	gsftp/
|	|	|	|	|	main.go
|	|	|	vendor/
|	|	|	|	github.com/pkg/sftp/
|	|	|	|	golang.org/x/crypto/ssh/
|	|	|	|	|	agent/
```

**Comment by myleshorton** [comment]
Interesting. I'm certainly not opposed to just waiting to see how this all
plays out in Go 1.5.

On Thu, Jun 11, 2015 at 12:15 PM, Ulysses Aalto <notifications@github.com>
wrote:

> There is news affecting this issue:
> https://groups.google.com/forum/#!msg/golang-dev/74zjMON9glU/4lWCRDCRZg0J
>
> Apparently, experimental support for vendoring in Go 1.5 will have pretty
> much the same philosophy of Gb. There seems to be a difference regarding
> placement of the project source code:
>
> Gb:
>
> .
> ├── src
> │   └── github.com
> │       └── pkg
> │           └── sftp
> │               └── examples
> │                   ├── buffered-read-benchmark
> │                   ├── buffered-write-benchmark
> │                   ├── gsftp
> │                   ├── streaming-read-benchmark
> │                   └── streaming-write-benchmark
> └── vendor
>     └── src
>         ├── github.com
>         │   └── kr
>         │       └── fs
>         └── golang.org
>             └── x
>                 └── crypto
>                     ...
>
> go -vendor:
>
> $GOPATH
> |   src/
> |   |   github.com/constabulary/example-gsftp/
> |   |   |   cmd/
> |   |   |   |   gsftp/
> |   |   |   |   |   main.go
> |   |   |   vendor/
> |   |   |   |   github.com/pkg/sftp/
> |   |   |   |   golang.org/x/crypto/ssh/
> |   |   |   |   |   agent/
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/issues/2595#issuecomment-111211078>
> .
>



-- 
--
President
Brave New Software Project, Inc.
https://www.getlantern.org
pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89

**Comment by aranhoide** [comment]
+1

**Comment by uaalto** [comment]
I think that's the wise thing to do. We can test it as soon as there is a beta and then decide.

**Comment by myleshorton** [comment]
Perfect!

On Fri, Jun 12, 2015 at 1:23 AM, Ulysses Aalto <notifications@github.com>
wrote:

> I think that's the wise thing to do. We can test it as soon as there is a
> beta and then decide.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/issues/2595#issuecomment-111376526>
> .
>



-- 
--
President
Brave New Software Project, Inc.
https://www.getlantern.org
pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89

**Comment by uaalto** [comment]
Update on this topic: It seems that the go --vendor support is going to use Gb's structure. Let's see if it gets implemented any time soon.

**Comment by myleshorton** [comment]
Closing for now. Let's open a new ticket when we know what's happening with --vendor

**Comment by uaalto** [comment]
More news on this front: https://medium.com/@freeformz/go-1-5-s-vendor-experiment-fd3e830f52c3
